ambry-router/src/main/java/com.github.ambry.router/EncryptDecryptThread.java

```diff
@@ -0,0 +1,264 @@
+/*
+ * Copyright 2017 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.router;
+
+import com.github.ambry.commons.BlobId;
+import java.nio.ByteBuffer;
+import java.security.GeneralSecurityException;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Thread responsible for encrypting and decrypting blob content and per blob keys. Thread listens to a job queue
+ * for new jobs and processes it one by one.
+ */
+public class EncryptDecryptThread implements Runnable {
+  private final BlockingQueue<JobInfo> jobQueue;
+  private final CryptoService cryptoService;
+  private final KeyManagementService kms;
+  private final AtomicBoolean enabled;
+  private final Lock lock;
+  private CountDownLatch shutdownLatch;
+
+  private static final Logger logger = LoggerFactory.getLogger(EncryptDecryptThread.class);
+
+  /**
+   * Instantiates {@link EncryptDecryptThread} with the job queue, {@link CryptoService} and {@link KeyManagementService}
+   * @param jobQueue the {@link BlockingQueue} that contains the jobs
+   * @param cryptoService the {@link CryptoService} to use to encrypt or decrypt
+   * @param kms the {@link KeyManagementService} to fetch keys from
+   */
+  EncryptDecryptThread(BlockingQueue<JobInfo> jobQueue, CryptoService cryptoService, KeyManagementService kms) {
+    this.jobQueue = jobQueue;
+    this.cryptoService = cryptoService;
+    this.kms = kms;
+    enabled = new AtomicBoolean(false);
+    lock = new ReentrantLock();
+  }
+
+  @Override
+  public void run() {
+    enabled.set(true);
+    shutdownLatch = new CountDownLatch(1);
+    try {
+      while (enabled.get()) {
+        lock.lock();
+        try {
+          JobInfo jobInfo = jobQueue.poll(10, TimeUnit.MILLISECONDS);
+          if (jobInfo != null) {
+            jobInfo.doOperation(cryptoService, kms);
+          }
+        } catch (InterruptedException e) {
+          logger.error("InterruptedException thrown in EncryptDecryptThread");
+        } finally {
+          lock.unlock();
+        }
+      }
+    } finally {
+      shutdownLatch.countDown();
+    }
+  }
+
+  /**
+   * Shuts down the thread by closing all pending jobs
+   */
+  void close() {
+    lock.lock();
```

Comment:
do you need this lock here? `run()` would just finish the job which it already removed from the queue and `closePendingJobs()` could just drain whatever is still on the queue.